### PR TITLE
Part1

### PR DESCRIPTION
**Pull request by remingtonchan**
Made errors go away.
Created proper constructors for Character derived types.
Implemented necessary virtual functions.
Implemented AttackItem class.
Fixed issues in Utility.
Made use of functions in Utility.
Added parameter for boosting to bypass printing out.
Addressed some warnings

### CODE REVIEW

File: Character.h [hunk 1]
```diff
@@ -35,32 +35,35 @@ struct Character
...
     }
     
-    void boostAttackDamage( int amount )
+    void boostAttackDamage( int amount, bool bypassCout = false )
     {
         attackDamage += amount;
-        std::cout << getName() << "'s attack damage level has been boosted to " << attackDamage << std::endl;
+        if (!bypassCout)
+            std::cout << getName() << "'s attack damage level has been boosted to " << attackDamage << std::endl;
     }
```
Comment (matkatmusic) [inline Character.h:57]: Do not modify these functions.
- [ ] revert

File: Character.cpp [hunk 2]
```diff
@@ -98,19 +99,38 @@ void Character::attackInternal(Character& other)
...
+        if (initialAttackDamage != nullptr)
+            attackDamage = *initialAttackDamage > attackDamage ? *initialAttackDamage : attackDamage;
+
+        boostHitPoints(hitPoints * 0.1f, true);
+        boostArmor(armor * 0.1f, true);
+        boostAttackDamage(attackDamage * 0.1f, true);
+
+        initialHitPoints.reset( new int(hitPoints) );
+        initialArmorLevel.reset( new int( armor) );
+        initialAttackDamage.reset( new int( attackDamage) );
```
Comment (matkatmusic) [inline Character.cpp:121]: a few things here require revision
- your constructor ensures that `initial____` is always valid and does not need to be checked if its a nullptr
- You are doing the same thing 3 times with 3 different pairs of variables. write a free function that updates each pair of ints accordingly.   hint: use reference arguments
- do not `reset` your `initial___` objects.   dereference and assign.
- [ ] revise

File: Character.cpp [hunk 3]
```diff
@@ -98,19 +99,38 @@ void Character::attackInternal(Character& other)
...
 {
     std::cout << getName() << "'s stats: " << std::endl;
-    assert(false);
+    // assert(false);
     /*
     make your getStats() use a function from the Utility.h
     */
-    std::cout << getStats(); 
+    // std::cout << getStats(); 
+    std::cout << getCharacterStats(this);
```
Comment (matkatmusic) [inline Character.cpp:133]: Your change here is unneeded.
You have to implement `getStats()` in each derived class.
- [ ] revert

File: Dragon.cpp [hunk 4]
```diff
@@ -18,5 +23,13 @@ void Dragon::attack(Character &other)
...
+    return name;
+}
 
-//Dragon::getStats
\ No newline at end of file
+//Dragon::getStats
+std::string Dragon::getStats()
+{
+    return "";
+}
```
Comment (matkatmusic) [inline Dragon.cpp:35]:
```
    /*
    make your getStats() use a function from the Utility.h
    */
```
There is a reason this function is pure virtual in the base class.
- [ ] revise

File: DragonSlayer.cpp [hunk 5]
```diff
@@ -25,4 +45,8 @@ void DragonSlayer::attack(Character& other)
         
 }
 
-//DragonSlayer::getStats
\ No newline at end of file
+//DragonSlayer::getStats
+std::string DragonSlayer::getStats()
+{
+    return "";
+}
```
Comment (matkatmusic) [inline DragonSlayer.cpp:52]:
```
    /*
    make your getStats() use a function from the Utility.h
    */
```
- [ ] revise

File: DragonSlayer.h [hunk 6]
```diff
@@ -1,8 +1,16 @@
...
+#include "AttackItem.h"
+
 struct DragonSlayer : Character
 {
+    DragonSlayer(std::string name_, int hpVal, int armorVal);
+    const std::string& getName() override;
+    std::string getStats() override;
     void attack(Character& other) override;
+
+    AttackItem* attackItem;
```
Comment (matkatmusic) [inline DragonSlayer.h:13]: Do not use raw pointers to manage objects.
use `std::unique_ptr` to manage heap-allocated objects
- [ ] revise

File: Dwarf.cpp [hunk 7]
```diff
@@ -1,7 +1,22 @@
...
+    return name;
+}
 
-//Dwarf::getStats
\ No newline at end of file
+//Dwarf::getStats
+std::string Dwarf::getStats()
+{
+    return "";
+}
```
Comment (matkatmusic) [inline Dwarf.cpp:22]:
```
    /*
    make your getStats() use a function from the Utility.h
    */
```
- [ ] revise

File: Paladin.cpp [hunk 8]
```diff
@@ -1,7 +1,22 @@
...
+    return name;
+}
 
-// Paladin::getStats
\ No newline at end of file
+// Paladin::getStats
+std::string Paladin::getStats()
+{
+    return "";
+}
```
Comment (matkatmusic) [inline Paladin.cpp:22]:
```
    /*
    make your getStats() use a function from the Utility.h
    */
```
- [ ] revise

File: Utility.cpp [hunk 9]
```diff
@@ -43,66 +49,68 @@ std::string getCharacterStats(Character* ch)
...
-        ch->boostAttackDamage(item->getBoost() * 1.33);
+        ch1->boostAttackDamage(item->getBoost() * 1.33);
     }
-    else if( auto* ch = dynamic_cast<DragonSlayer*>(character))
+    else if( auto* ch2 = dynamic_cast<DragonSlayer*>(character))
     {
-        assert(false);
+        // assert(false);
+
+        ch2->boostAttackDamage(ch2->getAttackDamage() * item->getBoost() * 10);
```
Comment (matkatmusic) [inline Utility.cpp:106]: There is nothing in the instructions that specify that you should multiply the attack damage and the boost by 10.

```
DragonSlayers get a 10x boost when attacking dragons, from their attack item.
so their attack item should boost their attack damage by a factor of 10
```
Look at the constructor arguments to `Item` and adjust your AttackItem constructor accordingly.
- [ ] revise

File: DragonSlayer.cpp [hunk 10]
```diff
@@ -1,16 +1,36 @@
...
     if( auto* dragon = dynamic_cast<Dragon*>(&other) )
     {
-        assert(false);
+        // assert(false);
+
+        if (attackItem != nullptr)
+        {
+            attackItem->use(this);
+            delete attackItem;
+            attackItem = nullptr;
```
Comment (matkatmusic) [inline DragonSlayer.cpp:32]:
```
//look in the Character class for how the other item types are reset after use.
```
- [ ] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]
Something for you to learn is when it is ok to modify existing code and when you should leave existing code alone. 
A lot of existing code was modified that should not have been modified.

File: Character.cpp (hunk 2 above)
Comment (remingtonchan) [reply Character.cpp:121]: Quick question here, do I abandon using the boost functions in the free function? (effectively applying the 10% on my own)?

File: DragonSlayer.h [hunk 11]
```diff
@@ -9,8 +9,7 @@ struct DragonSlayer : Character
     const std::string& getName() override;
     std::string getStats() override;
     void attack(Character& other) override;
-
-    AttackItem* attackItem;
+    std::unique_ptr<AttackItem> attackItem {new AttackItem()};
```
Comment (matkatmusic) [inline DragonSlayer.h:12]: This should be a private member variable.
No one outside of `DragonSlayer` uses it.
- [ ] revise

File: Character.cpp [hunk 12]
```diff
@@ -6,6 +6,13 @@
 #include "DefensiveItem.h"
 #include "HelpfulItem.h"
 
+void levelUpStat(int& statToLevel, int& initialStatVal) 
+{
+    statToLevel = initialStatVal > statToLevel ? initialStatVal : statToLevel;
```
Comment (matkatmusic) [inline Character.cpp:11]: Don't use ternary expressions unless both expressions change the value of the thing on the left of `=`
a simple `if( a > b )` is good enough here.
- [ ] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**